x-pack/packages/observability/observability_utils/observability_utils_server/entities/get_log_patterns.ts

```diff
@@ -0,0 +1,378 @@
+import { partition } from 'lodash';
+import { ObservabilityElasticsearchClient } from '../es/client/create_observability_es_client';
+import { kqlQuery } from '../es/queries/kql_query';
+import { rangeQuery } from '../es/queries/range_query';
+
+interface FieldPatternResultBase {
+  field: string;
+  count: number;
+  pattern: string;
+  regex: string;
+  sample: string;
+  firstOccurrence: string;
+  lastOccurrence: string;
+  highlight: Record<string, string[]>;
+  metadata: Record<string, unknown[]>;
+}
+
+interface FieldPatternResultChanges {
+  timeseries: Array<{ x: number; y: number }>;
+  change: {
+    significance: 'high' | 'medium' | 'low' | null;
+    type: ChangePointType;
+    change_point?: number;
+    r_value?: number;
+    trend?: string;
+    p_value?: number;
+  };
+}
+
+export type FieldPatternResult<TChanges extends boolean | undefined = undefined> =
+  FieldPatternResultBase & (TChanges extends true ? FieldPatternResultChanges : {});
+
+export type FieldPatternResultWithChanges = FieldPatternResult<true>;
+
+interface CategorizeTextOptions {
+  query: QueryDslQueryContainer;
+  metadata: string[];
+  esClient: ObservabilityElasticsearchClient;
+  samplingProbability: number;
+  fields: string[];
+  index: string | string[];
+  useMlStandardTokenizer: boolean;
+  size: number;
+  start: number;
+  end: number;
+}
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type CategorizeTextSubAggregations = {
+  sample: { top_hits: AggregationsTopHitsAggregation };
+  minTimestamp: { min: AggregationsMinAggregation };
+  maxTimestamp: { max: AggregationsMaxAggregation };
+};
+
+export async function runCategorizeTextAggregation<
+  TChanges extends boolean | undefined = undefined
+>(
+  options: CategorizeTextOptions & { changes?: TChanges }
+): Promise<Array<FieldPatternResult<TChanges>>>;
+
+export async function runCategorizeTextAggregation({
+  esClient,
+  fields,
+  metadata,
+  index,
+  query,
+  samplingProbability,
+  useMlStandardTokenizer,
+  changes,
+  size,
+  start,
+  end,
+}: CategorizeTextOptions & { changes?: boolean }): Promise<Array<FieldPatternResult<boolean>>> {
+  const aggs = Object.fromEntries(
+    fields.map(
+      (
+        field
+      ): [
+        string,
+        {
+          categorize_text: AggregationsCategorizeTextAggregation;
+          aggs: CategorizeTextSubAggregations &
+            (
+              | {}
+              | {
+                  timeseries: { date_histogram: AggregationsDateHistogramAggregation };
+                  changes: { change_point: { buckets_path: string } };
+                }
+            );
+        }
+      ] => [
+        field,
+        {
+          categorize_text: {
+            field,
+            min_doc_count: 1,
+            size,
+            categorization_analyzer: useMlStandardTokenizer
+              ? {
+                  tokenizer: 'ml_standard',
```

Comment:
I suggested this a few times, the approach here is as follows:

- use random sampling + standard tokenizer to get categories that are appearing most often
- exclude these patterns in an unsampled second request (which is possible by using the standard tokenizer) to get patterns that appear less often

Random sampling is disabled for now because of a [bug](https://github.com/elastic/elasticsearch/issues/110134), but I'll re-enable that and use a terms agg on _id instead of a top_hits agg, and then get the documents in a follow-up mget request. 

FWIW, I think we're better off talking about these patterns as "less frequent" or "low volume" - when I think about "rare" I think about the ML function, where it means that a pattern appears that is "less rare" than before, ie, it is new, or only appearing slightly more than before. ie, with your strategy, we are not detecting a change, we are simply detecting low volume. With the ML function, we _are_ detecting a change. I think "low frequency" is less useful than "rare'.